File: src/whirlwind/graph/edge_list.cpp
#include <string>

#include <nanobind/nanobind.h>
#include <nanobind/stl/pair.h>

#include <whirlwind/common/stddef.hpp>
#include <whirlwind/graph/csr_graph.hpp>
#include <whirlwind/graph/edge_list.hpp>
#include <whirlwind/graph/graph_traits.hpp>

#include "sequence.hpp"

namespace whirlwind::bindings {

namespace nb = nanobind;
using namespace nb::literals;

inline constexpr auto edge_list_add_edge_doc = R"(
    Add a new edge.

    Parameters
    ----------
    tail : Vertex
        The tail vertex of the edge.
    head : Vertex
        The head vertex of the edge.

    Returns
    -------
    EdgeList
        A reference to the `EdgeList` object.
    )";

template<class Vertex, template<class> class Container>
void
edge_list_attrs_and_methods(nb::class_<EdgeList<Vertex, Container>>& cls)
{
    using Class = EdgeList<Vertex, Container>;

    // Constructors.
    cls.def(nb::init<>());

    // Attributes & properties.
    cls.def_prop_ro("size", &Class::size);

    // Methods.
    cls.def("add_edge", &Class::add_edge, nb::rv_policy::reference_internal, "tail"_a,
            "head"_a, edge_list_add_edge_doc);
    cls.def("clear", &Class::clear, "Erase all edges from the container.");
    cls.def("empty", &Class::empty, "Check if the container is empty.");

    common_sequence_attrs_and_methods(cls);
}

void
edge_list(nb::module_& m)
{
    using Vertex = typename GraphTraits<CSRGraph<>>::vertex_type;
    auto edge_list = nb::class_<EdgeList<Vertex>>(m, "EdgeList");

    edge_list_attrs_and_methods(edge_list);
}

} // namespace whirlwind::bindings

File: src/whirlwind/graph/_edge_list.py
from collections.abc import Iterator

from . import _lib

__all__ = [
    "EdgeList",
]


class EdgeList:
    """A sequence of (tail,head) vertex pairs."""

    Vertex = int

    def __init__(self):  # type: ignore[no-untyped-def]
        self._impl = _lib.EdgeList()

    @property
    def size(self) -> int:
        """The total number of edges."""
        return self._impl.size

    def add_edge(self, tail: Vertex, head: Vertex) -> None:
        """
        Add a new edge.

        Parameters
        ----------
        tail : Vertex
            The tail vertex of the edge.
        head : Vertex
            The head vertex of the edge.
        """
        self._impl.add_edge(tail, head)

    def clear(self) -> None:
        """Erase all edges from the sequence."""
        self._impl.clear()

    def empty(self) -> bool:
        """Check if the container is empty."""
        return self._impl.empty()

    def __len__(self) -> int:
        return len(self._impl)

    def __getitem__(self, /, key: int) -> tuple[Vertex, Vertex]:
        return self._impl[key]

    def __iter__(self) -> Iterator[tuple[Vertex, Vertex]]:
        return iter(self._impl)
